### PR TITLE
Remove recusive memberlist locking

### PR DESCRIPTION
**Pull request by baumanj**
Since RwLocks can't be both fair and recursive (and recursive locking is very hard to reason about), let's eliminate the recursive locking of the locks in `MemberList`. See the individual commit messages for more explanation.

See https://github.com/habitat-sh/habitat/issues/6435

### CODE REVIEW

File: components/butterfly/src/server/mod.rs [hunk 1]
```diff
@@ -1141,22 +1164,6 @@ impl Server {
...
-        strukt.serialize_field("service", &self.service_store)?;
-        strukt.serialize_field("service_config", &self.service_config_store)?;
-        strukt.serialize_field("service_file", &self.service_file_store)?;
-        strukt.serialize_field("election", &self.election_store)?;
-        strukt.serialize_field("election_update", &self.update_store)?;
-        strukt.serialize_field("departure", &self.departure_store)?;
-        strukt.end()
-    }
-}
-
```
Comment (raskchanky) [inline components/butterfly/src/server/mod.rs]: Why was this removed?

Comment (baumanj) [reply components/butterfly/src/server/mod.rs]: It didn't appear that it was used. Since `Server` doesn't derive `Serialize` and removing this `impl Serialize for Server` didn't cause any complication errors, I concluded all serialization needs were being handled by `ServerProxy`. Removing `impl<'a> Serialize for ServerProxy<'a>` *does* break the build. Did I miss something?

Comment (raskchanky) [reply components/butterfly/src/server/mod.rs]: Ah, that's right.  I forgot about `ServerProxy`.
